### ISSUE TITLE
[feature-request] rewriting propagators as a differentiable pyTorch function and optimize with torch.optim

### ISSUE THREAD

**Issue by alvinccli** [computed-axial-lithography/VAMToolbox#20]
**Flexibility and performance improvements**

The current implementation of projection optimization consists of two layers.
The optimization layer (\optimizers\*.py) is implemented with numpy on CPU for compatibility with all existing propagators prior to v2.0.0.
The propagations (\projector\*.py) themselves are implemented with CUDA on GPU when possible (such as those from Astra toolbox and those explicitly written in pyTorch).

Future implementation could be doing everything on CUDA on GPU for better performance and integration.
The first step is to write the optimizer with GPU computation toolbox (such as pyTorch) so that the volumetric data does not have to be copied back and forth from GPU to CPU memory.

Another step further, is to also implement propagators (such as pyTorchRayTrace.py and pyTorchAlgebraicPropagation.py) as pyTorch functions that support backward auto-differentiation (AD). This will allow the pyTorch AD toolchain to directly differentiate on any loss function that involve these propagators. This way, we can use any torch.optim (a collection of gradient-based stepping schemes) to optimize arbitrary loss functions written even outside of the toolbox by users. This opens up a wide range of possible combinations of stepping scheme and loss functions.  For example, user A could write a new metric expression in torch and ask torch.optim to optimize that expression.
( Currently, even for latest BCLP, the gradient is computed from an analytical expression which is manually derived, and the stepping scheme is written by ourselves. )

**Comment by alvinccli** [comment]
To prepare for resolution of this issue, I propose to rename the 'forward' and 'backward' functions in all propagator to 'forwardPropagation' and 'backwardPropagation'.

The current naming convention follow those in Computed Tomography and happens to be the direct opposite of neural network convention. pyTorch call the corresponding actions 'backward' and 'forward' respectively in their toolchain. The renaming is to avoid naming clash.
